c_src/membrane_vpx_plugin/vpx_encoder.c

```diff
@@ -0,0 +1,158 @@
+#include "vpx_encoder.h"
+
+// The following code is based on the simple_encoder example provided by libvpx
+// (https://github.com/webmproject/libvpx/blob/main/examples/simple_encoder.c)
+
+void handle_destroy_state(UnifexEnv *env, State *state) {
+  UNIFEX_UNUSED(env);
+
+  vpx_codec_destroy(&state->codec_context);
+}
+
+vpx_img_fmt_t translate_pixel_format(PixelFormat pixel_format) {
+  switch (pixel_format) {
+  case PIXEL_FORMAT_I420:
+    return VPX_IMG_FMT_I420;
+  case PIXEL_FORMAT_I422:
+    return VPX_IMG_FMT_I422;
+  case PIXEL_FORMAT_I444:
+    return VPX_IMG_FMT_I444;
+  case PIXEL_FORMAT_YV12:
+    return VPX_IMG_FMT_YV12;
+  case PIXEL_FORMAT_NV12:
+    return VPX_IMG_FMT_NV12;
+  default:
+    return VPX_IMG_FMT_I420;
+  }
+}
+
+UNIFEX_TERM create(
+    UnifexEnv *env,
+    Codec codec,
+    unsigned int width,
+    unsigned int height,
+    PixelFormat pixel_format,
+    unsigned int encoding_deadline
+) {
+  UNIFEX_TERM result;
+  State *state = unifex_alloc_state(env);
+  vpx_codec_enc_cfg_t config;
+
+  switch (codec) {
+  case CODEC_VP8:
+    state->codec_interface = vpx_codec_vp8_cx();
+    break;
+  case CODEC_VP9:
+    state->codec_interface = vpx_codec_vp9_cx();
+    break;
+  }
+  state->encoding_deadline = encoding_deadline;
+
+  if (vpx_codec_enc_config_default(state->codec_interface, &config, 0)) {
+    return result_error(
+        env, "Failed to get default codec config", create_result_error, NULL, state
+    );
+  }
+
+  config.g_h = height;
+  config.g_w = width;
+  config.g_timebase.num = 1;
+  config.g_timebase.den = 1000000000; // 1e9
+  config.g_error_resilient = 1;
+
+  if (vpx_codec_enc_init(&state->codec_context, state->codec_interface, &config, 0)) {
+    return result_error(env, "Failed to initialize encoder", create_result_error, NULL, state);
+  }
+  if (!vpx_img_alloc(&state->img, translate_pixel_format(pixel_format), width, height, 1)) {
+    return result_error(
+        env, "Failed to allocate image", create_result_error, &state->codec_context, state
+    );
+  }
+  result = create_result_ok(env, state);
+  unifex_release_state(env, state);
+  return result;
+}
+
+void get_image_from_raw_frame(vpx_image_t *img, UnifexPayload *raw_frame) {
+  convert_between_image_and_raw_frame(img, raw_frame, RAW_FRAME_TO_IMAGE);
+}
+
+void alloc_output_frame(
+    UnifexEnv *env, const vpx_codec_cx_pkt_t *packet, UnifexPayload **output_frame
+) {
+  *output_frame = unifex_alloc(sizeof(UnifexPayload));
+  unifex_payload_alloc(env, UNIFEX_PAYLOAD_BINARY, packet->data.frame.sz, *output_frame);
+}
+
+UNIFEX_TERM encode(UnifexEnv *env, vpx_image_t *img, vpx_codec_pts_t pts, State *state) {
+  vpx_codec_iter_t iter = NULL;
+  int flushing = (img == NULL), got_packets = 0;
+  const vpx_codec_cx_pkt_t *packet = NULL;
+
+  unsigned int frames_cnt = 0, allocated_frames = 1;
+  UnifexPayload **encoded_frames = unifex_alloc(allocated_frames * sizeof(*encoded_frames));
```

Comment:
Is this syntax even allowed? :O What are `encoded_frames` in `sizeof(*encoded_frames)`? Does it mean the same as `sizeof(UnifexPayload*)`? :D

Reply:
i took it from h264_ffmpeg_plugin

Reply:
Could you set it to  `UnifexPayload *` instead? I have some trust issues with this syntax :D